File: code/go/ch17/src/luago/compiler/codegen/code_gen.go
package codegen

import . "luago/binchunk"
import . "luago/compiler/ast"

func GenProto(chunk *Block) *Prototype {
	fd := &FuncDefExp{
		IsVararg: true,
		Block:    chunk,
	}

	fi := newFuncInfo(nil, fd)
	fi.addLocVar("_ENV")
	cgFuncDefExp(fi, fd, 0)
	return toProto(fi.subFuncs[0])
}

File: code/java/ch18/src/main/java/com/github/zxh0/luago/compiler/codegen/CodeGen.java
package com.github.zxh0.luago.compiler.codegen;

import com.github.zxh0.luago.binchunk.Prototype;
import com.github.zxh0.luago.compiler.ast.Block;
import com.github.zxh0.luago.compiler.ast.exps.FuncDefExp;

public class CodeGen {

    public static Prototype genProto(Block chunk) {
        FuncDefExp fd = new FuncDefExp();
        fd.setLastLine(chunk.getLastLine());
        fd.setIsVararg(true);
        fd.setBlock(chunk);

        FuncInfo fi = new FuncInfo(null, fd);
        fi.addLocVar("_ENV", 0);
        ExpProcessor.processFuncDefExp(fi, fd, 0);
        return Fi2Proto.toProto(fi.subFuncs.get(0));
    }

}
